docs/reference/configuration.md
# Configuration

Configuration is a key part of the Custom Pod Autoscaler, defining how to call user logic, alongside more fine tuned
configuration options that the end user might adjust; such as polling interval.

## How to provide configuration

### Configuration file in image

Example:

```yaml
evaluate:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/evaluate.py"
metric:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/metric.py"
runMode: "per-resource"
```

Configuration may be set by using a configuration file (default path `/config.yaml`), this can be baked into a Docker
image to provide standard configuration for the autoscaler - for example metric gathering and evaluation methods.

### Configuration passed as environment variables (defined in deployment YAML)

Example:

```yaml
  config:
    - name: interval
      value: "10000"
    - name: startTime
      value: "60000"
```

Configuration may be passed as environment variables; these can be customised at deploy time rather than baked in at
build time, so allow for more fine tuned customisation. The main way to define the environment variables should be
using the `config` YAML description, which allows for key-value pairs to define each configuration option and the value
it should have. All configuration options defined as key-value pairs in the `config` YAML are converted into environment
variables by the operator; this allows autoscalers to extend the configuration options and use this `config` YAML to
define configuration for the user logic.


> Note: Configuration set as an environment variable takes precedence over configuration set in a configuration file,
> this allows the configuration file to act possibly as a set of defaults that can be overridden at deploy time.

## Methods

Example:

```yaml
metric:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/metric.py"
```

A method defines a hook for calling user logic, with a timeout to handle user
logic that hangs.

### type

- `shell` = call the user logic through a shell command, with the relevant
  information piped through to the command. The user logic communicates back
  with the autoscaler through exit codes and standard error and out. A non zero
  exit code tells the autoscaler that the user logic has failed; and the
  autoscaler will read in standard error and log it for debug purposes. If no
  error occurs, the autoscaler may read in the standard out and use it, e.g. for
  metric gathering.
- `http` = make an HTTP request to retrieve the results, with the relevant
  information passed through via a parameter (either body or query). The user
  logic communicates back with the autoscaler through HTTP status codes and the
  response body. A status code that is defined as not successful (configured as
  part of the method) tells the autoscaler that the request has failed; while a
  successful response code indicates success.

Defines the type of the method.

### timeout

Defines how long the autoscaler should wait for the user logic to finish, if it exceeds this time it will assume the
operation has failed and provide a timeout error.

### shell

Defines a shell method, which is a simple string with the shell command to execute. Shell commands are configured with
entrypoints and commands to execute, with values piped in through standard in. See the [methods section for more
details](../user-guide/methods.md#shell)

### http

Defines an http method, which configures how an HTTP request is made. See the [methods section for more
details](../user-guide/methods.md#http)

## configPath

```yaml
config:
  - name: configPath
    value: "/config.yaml"
```

Default: `/config.yaml`

This defines the path to the configuration file. Should only be defined as an environment variable (through deployment
YAML), as defining the path to the configuration file inside the configuration file does not make sense.

## interval

Example:

```yaml
interval: 15000
```

Default value: `15000`

This defines in milliseconds how frequently the autoscaler should run. The autoscaler will run, then wait this many
milliseconds before running again.

## evaluate

Example:

```yaml
evaluate:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/evaluate.py"
```

No default value, required to be set.

This defines the evaluation logic that should be run, and how it should be triggered.
[This is a `method`, see methods section for full configuration options of a method](#methods).

## metric

Example:

```yaml
metric:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/metric.py"
```

No default value, required to be set.

This defines the metric logic that should be run, and how it should be triggered.
[This is a `method`, see methods section for full configuration options of a method](#methods).

## namespace

Example:

```yaml
namespace: "default"
```

Default value: `default`

Defines the namespace to look in for the resource being managed.

## minReplicas

Example:

```yaml
minReplicas: 1
```

Default value: `1`

Defines the minimum number of replicas allowed, resource won't be scaled below this value. If set to anything but `0`
scaling to zero is disabled, and a replica count of `0` will be treated as autoscaling disabled. If set to `0` scaling
to `0` is enabled.

## maxReplicas

Example:

```yaml
maxReplicas: 10
```

Default value: `10`

Defines the maximum number of replicas allowed, resource won't be scaled above this value.

## runMode

Example:

```yaml
runMode: "per-pod"
```

Default value: `per-pod`

- `per-pod` = runs metric gathering per pod, individually running the user logic for each pod in the resource being
managed, with the pod information provided to the user logic.
- `per-resource` = runs metric gathering per resource, running the user logic for only the resource being managed, with
the resource information provided to the user logic.

Defines how the autoscaler runs the metric gathering user logic, changing the values that are provided to the metric
gathering user logic and changing how frequently the metric gathering user logic is called.

## startTime

Example:

```yaml
startTime: 1
```

Default value: `1`

This defines in milliseconds a starting point for the scaler, with the scaler running as if it started at the time
provided. Allows specifying that the autoscaler must start on a multiple of the interval from the start time. For
example, a startTime of `60000` would result in the autoscaler starting at the next full minute. The default value will
start the autoscaler after a single millisecond, close to instantly.

> Note: The scaling will not actually start until after one interval has passed.

## logVerbosity

Example:

```yaml
logVerbosity: 0
```

Default value: `0`

This defines the verbosity of the logging, allowing for debugging errors/issues. Logging will occur for all values
ABOVE and including the verbosity level set.

Log levels:

* `0` - normal.
* `1` - verbose.
* `2` - more verbose around high level logic such as autoscaling/rest api.
* `3` - more verbose around lower level logic such as metric gathering and evaluation.

## downscaleStabilization

Example:

```yaml
downscaleStabilization: 200
```

Default value: `0`

This defines in seconds the length of the downscale stabilization window; based on [the Horizontal Pod Autoscaler
downscale
stabilization](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-cooldown-delay).
Downscale stabilization works by recording all evaluations over the window specified and picking out the maximum target
replicas from these evaluations. This results in a more smoothed downscaling and a cooldown, which can reduce the effect
of thrashing.

## apiConfig

Example:

```yaml
apiConfig:
  enabled: true
  useHTTPS: true
  port: 80
  host: "0.0.0.0"
  certFile: "cert.crt"
  keyFile: "key.key"
```

Default value:

```yaml
apiConfig:
  enabled: true
  useHTTPS: false
  port: 5000
  host: "0.0.0.0"
  certFile: ""
  keyFile: ""
```

This sets configuration options for the Custom Pod Autoscaler API, allowing enabling and disabling the API and
specifying how the API should be exposed.

### enabled

Boolean value to enable (`true`) or disable (`false`) the API

### useHTTPS

Boolean value to enable (`true`) or disable (`false`) HTTPS.

### port

Integer value defining the port to expose the API on.

### host

String value defining the host to expose the API on.

### certFile

String value defining the path to the [certificate](https://golang.org/pkg/net/http/#ListenAndServeTLS) to use for
HTTPS.

### keyFile

String value defining the path to the [private key](https://golang.org/pkg/net/http/#ListenAndServeTLS) to use for
HTTPS.

## preMetric

Example:

```yaml
preMetric:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/pre-metric.py"
```

No default value, if not set it is not executed.

This defines a pre-metric hook, and how it should be triggered.
The pre-metric hook is run before metric gathering occurs, it is provided with either the resource being managed or the
pods being managed (depending on the run mode) as JSON.

Example of JSON provided to this hook:

```json
{
  "resource": {
    "kind": "Deployment",
    "apiVersion": "apps/v1",
    "metadata": {
      "name": "hello-kubernetes",
      "namespace": "default",
    },
    ...
  },
  "runType": "scaler"
}
```

[This is a `method`](#methods) that is running as part of a [`hook`](../user-guide/hooks.md).

## postMetric

Example:

```yaml
postMetric:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/post-metric.py"
```

No default value, if not set it is not executed.

This defines a post-metric hook, and how it should be triggered.
The post-metric hook is run after metric gathering occurs, it is provided with either the resource being managed or the
pods being managed (depending on the run mode) alongside the metric gathering results as JSON.

Example of JSON provided to this hook:
```json
{
  "resource": {
    "kind": "Deployment",
    "apiVersion": "apps/v1",
    "metadata": {
      "name": "hello-kubernetes",
      "namespace": "default",
    },
    ...
  },
  "metrics": [
    {
      "resource": "hello-kubernetes",
      "value": "3"
    }
  ],
  "runType": "scaler"
}
```

[This is a `method`](#methods) that is running as part of a [`hook`](../user-guide/hooks.md).

## preEvaluate

Example:

```yaml
preEvaluate:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/pre-evaluate.py"
```

No default value, if not set it is not executed.

This defines a pre-evaluate hook, and how it should be triggered.
The pre-evaluate hook is run before evaluation occurs, it is provided with the full resource metrics as JSON.

Example of JSON provided to this hook:
```json
{
  "metrics": [
    {
      "resource": "hello-kubernetes",
      "value": "3"
    }
  ],
  "resource": {
    "kind": "Deployment",
    "apiVersion": "apps/v1",
    "metadata": {
      "name": "hello-kubernetes",
      "namespace": "default",
    },
    ...
  },
  "runType": "scaler"
}
```

[This is a `method`](#methods) that is running as part of a [`hook`](../user-guide/hooks.md).

## postEvaluate

Example:

```yaml
postEvaluate:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/post-evaluate.py"
```

No default value, if not set it is not executed.

This defines a post-evaluate hook, and how it should be triggered.
The post-evaluate hook is run after evaluation occurs, it is provided with the full resource metrics alongside the
evaluation that has been calculated as JSON.

Example of JSON provided to this hook:
```json
{
  "metrics": [
    {
      "resource": "hello-kubernetes",
      "value": "3"
    }
  ],
  "resource": {
    "kind": "Deployment",
    "apiVersion": "apps/v1",
    "metadata": {
      "name": "hello-kubernetes",
      "namespace": "default",
    },
    ...
  },
  "evaluation": {
    "targetReplicas": 3
  },
  "runType": "scaler"
}
```

[This is a `method`](#methods) that is running as part of a [`hook`](../user-guide/hooks.md).

## preScale

Example:

```yaml
preScale:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/pre-scale.py"
```

No default value, if not set it is not executed.

This defines a pre-scaling hook, and how it should be triggered.
This hook is run even if autoscaling is disabled for the resource (replicas set to `0`).
The pre-scale hook is run before a scaling decision is made, it is provided with min and max replicas, current replicas,
target replicas, and resource being scaled as JSON.

Example of JSON provided to this hook:

```json
{
  "evaluation": {
    "targetReplicas": 6
  },
  "resource": {
    "kind": "Deployment",
    "apiVersion": "apps/v1",
    "metadata": {
      "name": "hello-kubernetes",
      "namespace": "default",
    },
    ...
  },
  "scaleTargetRef": {
    "kind": "Deployment",
    "name": "hello-kubernetes",
    "apiVersion": "apps/v1"
  },
  "namespace": "default",
  "minReplicas": 1,
  "maxReplicas": 10,
  "targetReplicas": 0,
  "runType": "scaler"
}
```

[This is a `method`](#methods) that is running as part of a [`hook`](../user-guide/hooks.md).

## postScale

Example:

```yaml
postScale:
  type: "shell"
  timeout: 2500
  shell:
    entrypoint: "python"
    command:
      - "/post-scale.py"
```

No default value, if not set it is not executed.

This defines a post-scaling hook, and how it should be triggered.
This hook is only run if scaling is successful, and is not run if autoscaling is disabled for the resource (replicas
set to `0`).

The post-scale hook is run after a scaling decision is made and effected, it is provided with min and max replicas,
current replicas, target replicas, and resource being scaled as JSON.

Example of JSON provided to this hook:

```json
{
  "evaluation": {
    "targetReplicas": 6
  },
  "resource": {
    "kind": "Deployment",
    "apiVersion": "apps/v1",
    "metadata": {
      "name": "hello-kubernetes",
      "namespace": "default",
    },
    ...
  },
  "scaleTargetRef": {
    "kind": "Deployment",
    "name": "hello-kubernetes",
    "apiVersion": "apps/v1"
  },
  "namespace": "default",
  "minReplicas": 1,
  "maxReplicas": 10,
  "targetReplicas": 0,
  "runType": "scaler"
}
```

[This is a `method`](#methods) that is running as part of a [`hook`](../user-guide/hooks.md).

## kubernetesMetricSpecs

Example:

```yaml
kubernetesMetricSpecs:
  - type: Resource
    resource:
      name: cpu
      target:
        type: Utilization
  - type: Resource
    resource:
      name: memory
      target:
        type: AverageValue
```

No default value, if not set the Kubernetes metrics server will not be queried.

This defines a list of the Kubernetes metrics to fetch from the Kubernetes metrics server. These metrics are then
serialized and then fed into the metric gathering stage of the Custom Pod Autoscaler pipeline.

This option supports any Kubernetes metrics that can be defined as part of a Kubernetes Horizontal Pod Autoscaler:

> Note: Since the Custom Pod Autoscaler leaves evaluation of the metrics to the user logic, there is no need to provide
> target values for defining metric specs, so the `averageValue`, `value`, and `averageUtilization` values can be
> left out

For more information about the Kubernetes metrics and how to define which Kubernetes metrics the Custom Pod Autoscaler
needs see the Kubernetes documentation here:

<https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/#autoscaling-on-multiple-metrics-and-custom-metrics>


The metrics will be serialized in the following form:

```json
...
"kubernetesMetrics": [
  {
    "current_replicas": <current replicas>,
    "spec": <the metric spec used to generate the metrics in JSON>,
    "resource": <if it is a resource metric, this will be populated, otherwise it will not be set>,
    "pods": <if it is a pods metric, this will be populated, otherwise it will not be set>,
    "object": <if it is a object metric, this will be populated, otherwise it will not be set>,
    "external": <if it is a external metric, this will be populated, otherwise it will not be set>
  }
]
...
```

Each type of metric has a different JSON structure which must be processed differently, since they return different
sets of data.

### resource

Example of a `resource` metric spec:

```yaml
- type: Resource
  resource:
    name: cpu
    target:
      type: Utilization
```

The `resource` metric will be serialized into the following form:

```json
"resource": {
  "pod_metrics_info": <list of pod metrics>,
  "requests": <map of pod names to their request value of the resource being queried (e.g. CPU requests)>,
  "ready_pod_count": <number of ready pods>,
  "ignored_pods": <a set of ignored pod names>,
  "missing_pods": <a set of missing pod names>,
  "total_pods": <number of ready pods>,
  "timestamp": <timestamp>
}
```

Example of the serialised JSON Kubernetes metrics provided to the metric gathering stage if metric specs are provided:

```json
{
  "resource": {...},
  "runType": "scaler",
  "kubernetesMetrics": [
    {
      "current_replicas": 1,
      "spec": {
        "type": "Resource",
        "resource": {
          "name": "cpu",
          "target": {
            "type": "Utilization"
          }
        }
      },
      "resource": {
        "pod_metrics_info": {
          "flask-metric-697794dd85-bsttm": {
            "Timestamp": "2021-04-05T18:10:10Z",
            "Window": 30000000000,
            "Value": 4
          }
        },
        "requests": {
          "flask-metric-697794dd85-bsttm": 200
        },
        "ready_pod_count": 1,
        "ignored_pods": {},
        "missing_pods": {},
        "total_pods": 1,
        "timestamp": "2021-04-05T18:10:10Z"
      }
    }
  ]
}
```

### pods

Example of a `pods` metric spec:

```yaml
- type: Pods
  pods:
    metric:
      name: packets-per-second
    target:
      type: AverageValue
```

The `pods` metric will be serialized into the following form:

```json
"pods": {
  "pod_metrics_info": <list of pod metrics>,
  "ready_pod_count": <number of ready pods>,
  "ignored_pods": <a set of ignored pod names>,
  "missing_pods": <a set of missing pod names>,
  "total_pods": <number of ready pods>,
  "timestamp": <timestamp>
}
```

Example of the serialised JSON Kubernetes metrics provided to the metric gathering stage if metric specs are provided:

```json
{
  "resource": {...},
  "runType": "scaler",
  "kubernetesMetrics": [
    {
      "current_replicas": 1,
      "spec": {
        "type": "Pods",
        "pods": {
          "metric": {
            "name": "packets-per-second",
          },
          "target": {
            "type": "AverageValue"
          }
        }
      },
      "resource": {
        "pod_metrics_info": {
          "flask-metric-697794dd85-bsttm": {
            "Timestamp": "2021-04-05T18:10:10Z",
            "Window": 30000000000,
            "Value": 1000
          }
        },
        "ready_pod_count": 1,
        "ignored_pods": {},
        "missing_pods": {},
        "total_pods": 1,
        "timestamp": "2021-04-05T18:10:10Z"
      }
    }
  ]
}
```

### object

Example of an `object` metric spec:

```yaml
- type: Object
  object:
    metric:
      name: requests-per-second
    describedObject:
      apiVersion: networking.k8s.io/v1beta1
      kind: Ingress
      name: main-route
    target:
      type: Value
```

The `object` metric will be serialized into the following form:

```json
"object": {
  "current": <the current value of the metric (value or average value)>,
  "ready_pod_count": <number of ready pods>,
  "timestamp": <timestamp>
}
```

Example of the serialised JSON Kubernetes metrics provided to the metric gathering stage if metric specs are provided:

```json
{
  "resource": {...},
  "runType": "scaler",
  "kubernetesMetrics": [
    {
      "current_replicas": 1,
      "spec": {
        "type": "External",
        "metric": {
          "name": "queue_messages_ready",
          "selector": "queue=worker_tasks",
          "target": {
            "type": "AverageValue"
          }
        }
      },
      "external": {
        "current": {
          "average_value": 5
        },
        "ready_pod_count": 1,
        "timestamp": "2021-04-05T18:10:10Z"
      }
    }
  ]
}
```

### external

Example of an `external` metric spec:

```yaml
- type: External
  external:
    metric:
      name: queue_messages_ready
      selector: "queue=worker_tasks"
    target:
      type: AverageValue
```

The `external` metric will be serialized into the following form:

```json
"external": {
  "current": <the current value of the metric (value or average value)>,
  "ready_pod_count": <number of ready pods>,
  "timestamp": <timestamp>
}
```

Example of the serialised JSON Kubernetes metrics provided to the metric gathering stage if metric specs are provided:

```json
{
  "resource": {...},
  "runType": "scaler",
  "kubernetesMetrics": [
    {
      "current_replicas": 1,
      "spec": {
        "type": "External",
        "metric": {
          "name": "queue_messages_ready",
          "selector": "queue=worker_tasks",
          "target": {
            "type": "Value"
          }
        }
      },
      "external": {
        "current": {
          "value": 5
        },
        "ready_pod_count": 1,
        "timestamp": "2021-04-05T18:10:10Z"
      }
    }
  ]
}
```

## requireKubernetesMetrics

Example:

```yaml
requireKubernetesMetrics: true
```

Default value: `false`.

Flag that allows the Custom Pod Autoscaler to fail if it does not successfully query the Kubernetes metrics server.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic
Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [v2.4.0] - 2021-12-28
### Changed
- Error messages now include more context about when they occured, rather than simply outputting the minimum error
information.

## [v2.3.0] - 2021-12-27
### Added
- Support for `argoproj.io/v1alpha1` `Rollout` resource.
### Changed
- Add clarity between `autoscaling/v1` and `autoscaling/v2beta2` APIs, no breaking changes.
### Security
- Updated k8s client libraries to `v0.21.8` to resolve
[GHSA-f5f7-6478-qm6p](https://github.com/advisories/GHSA-f5f7-6478-qm6p).

## [v2.2.0] - 2021-08-14
### Added
- New `openjdk-11` image which includes Java 11 runtime.

## [v2.1.0] - 2021-06-16
### Added
- New `logStderr` option for `shell` methods, intended to help debugging by printing out the stderr of a successful
`shell` method execution - default value `false`.
### Changed
- Error returned when parsing invalid JSON for an evaluation improved to help debugging.

## [v2.0.0] - 2021-04-15
### Changed
- **BREAKING CHANGE** Project's Go code restructured, limited exposed packages. See [the migration
guide](./docs/user-guide/migrating-v1-to-v2.md) for full details.
- **BREAKING CHANGE** Pre-scaling hook moved to after the downscale stabilization value has been calculated.
- `k8smetric` package now exposed to allow easy Go dependency marshal/unmarshal of K8s metrics.
### Fixed
- `targetReplicas` value now set properly, set to the pre-stabilized target replica value.
### Added
- The Custom Pod Autoscaler version is now printed to the log on autoscaler startup.

## [v1.1.0] - 2021-04-08
### Added
- Feature to provide standard K8s metrics to the metric gathering stage of the autoscaler
  - Can now provide a list of Metric Specs (similar to the Horizontal Pod Autoscaler) to choose which metrics to
  include in the data sent to the metric gathering stage with the `kubernetesMetricSpecs` configuration option.
  - Can provide the `requireKubernetesMetrics` option to fail if the metrics server query fails.
  - Can provide `initialReadinessDelay` and `cpuInitializationPeriod` values for use when querying the metrics server.
### Changed
- Switched docs theme to material theme.

## [v1.0.1] - 2020-09-12
### Added
- Three new Python images:
  * `custompodautoscaler/python-3-6` tracks latest stable Python 3.6.x release.
  * `custompodautoscaler/python-3-7` tracks latest stable Python 3.7.x release.
  * `custompodautoscaler/python-3-8` tracks latest stable Python 3.8.x release.
### Changed
- The `custompodautoscaler/python` image now tracks the latest stable Python 3.x release.

## [v1.0.0] - 2020-07-19

## [v0.13.0] - 2020-07-18
### Added
- HTTP method, allows specifying an HTTP request to make as a method - for example querying an external API as part of
the metric gathering stage.
- Extra error checking for `shell` method, will no longer throw nil pointer if no shell configuration is provided, more
useful error is raised instead.

## [v0.12.0] - 2020-04-25
### Changed
- Support scaling to and from zero, matching misimplemented functionality from Horizontal Pod Autoscaler.

## [v0.11.0] - 2020-02-28
### Added
- Series of hooks for injecting user logic throughout the execution process.
  * `preMetric` - Runs before metric gathering, given metric gathering input.
  * `postMetric` - Runs after metric gathering, given metric gathering input and result.
  * `preEvaluate` - Runs before evaluation, given evaluation input.
  * `postEvaluate` - Runs after evaluation, given evaluation input and result.
  * `preScale` - Runs before scaling decision, given min and max replicas, current replicas, target replicas, and
  resource being scaled.
  * `postScale` - Runs before scaling decision, given min and max replicas, current replicas, target replicas, and
  resource being scaled.
- New `downscaleStabilization` option, based on [the Horizontal Pod Autoscaler downscale
  stabilization](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-cooldown-delay),
  operates by taking the maximum target replica count over the stabilization window.
### Changed
- Metrics from API now returns the entire resource definition as JSON rather than just the resource name.
- Changed JSON generated to be in `camelCase` rather than `snake_case` for consistency with the Kubernetes API.
  * Evaluation now uses `targetReplicas` over `target_replicas`.
  * ResourceMetric now uses `runType` over `run_type`.
  * Scale hook now provided with `minReplicas`, `maxReplicas`, `currentReplicas` and `targetReplicas` rather than their
  snakecase equivalents.
- Metric gathering and hooks have access to `dryRun` field, allowing them to determine if they are called as part of a
  dry run.
- Standardised input to metric gatherer, evaluator and scaler to take specs rather than lists of parameters, allowing
  easier serialisation for hooks.
- Endpoint `/api/v1/metrics` now accepts the optional `dry_run` parameter for marking metric gathering as in dry run
  mode.
- `ResourceMetrics` replaced with a list of `Metric` and a `Resource`.
- `/api/v1/metrics` now simply returns a list of `Metrics` rather than a `ResourceMetrics`.

### Removed
- `ResourceMetrics` struct removed as it was redundant.

## [v0.10.0] - 2020-01-22
### Added
- Set up API to be versioned, starting with `v1`.
- Can now manually trigger scaling through the API.
- Added extra `run_type` flag, `api_dry_run`, for evaluations through the API in `dry_run` mode.
- Added `apiConfig` to hold configuration for the REST API.
- Added extra configuration options within `apiConfig`.
  * `enabled` - allows enabling or disabling the API, default enabled (`true`).
  * `useHTTPS` - allows enabling or disabling HTTPS for the API, default off (`false`).

config/config.go
/*
Copyright 2021 The Custom Pod Autoscaler Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package config handles parsing YAML and environment variable input into a
// Custom Pod Autoscaler configuration file. Contains a set of defaults that
// can be overridden by provided YAML and env vars.
package config

import (
	"bytes"
	"reflect"
	"strconv"
	"strings"

	"github.com/jthomperoo/custom-pod-autoscaler/internal/measure"
	autoscaling "k8s.io/api/autoscaling/v1"
	"k8s.io/apimachinery/pkg/util/yaml"
)

const (
	// PerPodRunMode runs metric gathering per Pod, individually running the script for each Pod being managed
	// with the Pod information piped into the metric gathering script
	PerPodRunMode = "per-pod"
	// PerResourceRunMode runs metric gathering per Resource, running the script only once for the resource
	// being managed, with the resource information piped into the metric gathering script
	PerResourceRunMode = "per-resource"
)

const (
	defaultInterval                = 15000
	defaultNamespace               = "default"
	defaultMinReplicas             = 1
	defaultMaxReplicas             = 10
	defaultStartTime               = 1
	defaultRunMode                 = PerPodRunMode
	defaultLogVerbosity            = 0
	defaultDownscaleStabilization  = 0
	defaultCPUInitializationPeriod = 300
	defaultInitialReadinessDelay   = 30
)

const (
	defaultAPIEnabled = true
	defaultUseHTTPS   = false
	defaultHost       = "0.0.0.0"
	defaultPort       = 5000
	defaultCertFile   = ""
	defaultKeyFile    = ""
)

const jsonStructTag = "json"

// Config is the configuration options for the CPA
type Config struct {
	ScaleTargetRef           *autoscaling.CrossVersionObjectReference `json:"scaleTargetRef"`
	PreMetric                *Method                                  `json:"preMetric"`
	PostMetric               *Method                                  `json:"postMetric"`
	PreEvaluate              *Method                                  `json:"preEvaluate"`
	PostEvaluate             *Method                                  `json:"postEvaluate"`
	PreScale                 *Method                                  `json:"preScale"`
	PostScale                *Method                                  `json:"postScale"`
	Evaluate                 *Method                                  `json:"evaluate"`
	Metric                   *Method                                  `json:"metric"`
	Interval                 int                                      `json:"interval"`
	Namespace                string                                   `json:"namespace"`
	MinReplicas              int32                                    `json:"minReplicas"`
	MaxReplicas              int32                                    `json:"maxReplicas"`
	RunMode                  string                                   `json:"runMode"`
	StartTime                int64                                    `json:"startTime"`
	LogVerbosity             int32                                    `json:"logVerbosity"`
	DownscaleStabilization   int                                      `json:"downscaleStabilization"`
	APIConfig                *APIConfig                               `json:"apiConfig"`
	KubernetesMetricSpecs    []measure.MetricSpec                     `json:"kubernetesMetricSpecs"`
	RequireKubernetesMetrics bool                                     `json:"requireKubernetesMetrics"`
	InitialReadinessDelay    int64                                    `json:"initialReadinessDelay"`
	CPUInitializationPeriod  int64                                    `json:"cpuInitializationPeriod"`
}

// APIConfig is configuration options specifically for the API exposed by the CPA
type APIConfig struct {
	Enabled  bool   `json:"enabled"`
	UseHTTPS bool   `json:"useHTTPS"`
	Port     int    `json:"port"`
	Host     string `json:"host"`
	CertFile string `json:"certFile"`
	KeyFile  string `json:"keyFile"`
}

// Method describes a method for passing data/triggerering logic, such as through a shell
// command
type Method struct {
	Type    string `json:"type"`
	Timeout int    `json:"timeout"`
	Shell   *Shell `json:"shell"`
	HTTP    *HTTP  `json:"http"`
}

// Shell describes configuration options for a shell command method
type Shell struct {
	Command    []string `json:"command"`
	Entrypoint string   `json:"entrypoint"`
}

// HTTP describes configuration options for an HTTP request method
type HTTP struct {
	Method        string            `json:"method"`
	URL           string            `json:"url"`
	Headers       map[string]string `json:"headers,omitempty"`
	SuccessCodes  []int             `json:"successCodes"`
	ParameterMode string            `json:"parameterMode"`
}

// LoadConfig loads in the default configuration, then overrides it from the config file,
// then any env vars set.
func LoadConfig(configFileData []byte, envVars map[string]string) (*Config, error) {
	config := newDefaultConfig()
	err := loadFromBytes(configFileData, config)
	if err != nil {
		return nil, err
	}
	err = loadFromEnv(config, envVars)
	if err != nil {
		return nil, err
	}
	// Check API defaults

	return config, nil
}

func loadFromBytes(data []byte, config *Config) error {
	// If no bytes file data provided, skip trying to parse it
	if data == nil {
		return nil
	}
	err := yaml.NewYAMLOrJSONDecoder(bytes.NewReader(data), 10).Decode(config)
	if err != nil {
		return err
	}
	return nil
}

func loadFromEnv(config *Config, envVars map[string]string) error {
	// Get config Go types and values
	configTypes := reflect.TypeOf(config).Elem()
	configValues := reflect.ValueOf(config).Elem()

	// Iterate through each field in the config
	for i := 0; i < configTypes.NumField(); i++ {
		// Get each field's type and value
		fieldType := configTypes.Field(i)
		fieldValue := configValues.Field(i)

		// Extract JSON tag from the type, e.g `json:"example"` would return example
		tag := fieldType.Tag.Get(jsonStructTag)

		// Check if there is an environment variable provided with the same tag
		value, exists := envVars[tag]
		if !exists {
			continue
		}

		// Assign values using correct types
		if fieldValue.Kind() == reflect.String {
			fieldValue.SetString(value)
			continue
		}
		if fieldValue.Kind() == reflect.Int {
			intVal, err := strconv.ParseInt(value, 10, 64)
			if err != nil {
				return err
			}
			fieldValue.SetInt(intVal)
			continue
		}

		// If the type is not one of the primitives above, it must be in JSON or YAML form, so try to parse
		// it and set the value from the unmarshalled JSON or YAML value
		fieldRef := reflect.New(fieldType.Type)
		err := yaml.NewYAMLOrJSONDecoder(strings.NewReader(value), 10).Decode(fieldRef.Interface())
		if err != nil {
			return err
		}

		fieldValue.Set(fieldRef.Elem())
		continue
	}
	return nil
}

func newDefaultConfig() *Config {
	return &Config{
		Interval:               defaultInterval,
		Namespace:              defaultNamespace,
		MinReplicas:            defaultMinReplicas,
		MaxReplicas:            defaultMaxReplicas,
		StartTime:              defaultStartTime,
		RunMode:                defaultRunMode,
		DownscaleStabilization: defaultDownscaleStabilization,
		APIConfig: &APIConfig{
			Enabled:  defaultAPIEnabled,
			UseHTTPS: defaultUseHTTPS,
			Port:     defaultPort,
			Host:     defaultHost,
			CertFile: defaultCertFile,
			KeyFile:  defaultKeyFile,
		},
		KubernetesMetricSpecs:    nil,
		RequireKubernetesMetrics: false,
		InitialReadinessDelay:    defaultInitialReadinessDelay,
		CPUInitializationPeriod:  defaultCPUInitializationPeriod,
	}
}
